## Last name categorization in census analysis
#### Professor:  Melinda C. Miller
#### Name: Haizhou Li
#### Data: 2020-09

Used pandas and numpy to do the data cleaning, and match the names by using Fuzzy Name Matching Algorithms and imported this pakage whoswho

Pakage install (In google colab, change the exclamation point to question mark)

In [ ]:
! pip install whoswho

In [1]:
import pandas as pd
import numpy as np
from whoswho import who
import datetime

Import data sets

In [6]:
df = pd.read_csv("https://raw.githubusercontent.com/hydro9811/Name-Comparison-/main/Copy%20of%201900%20AI%20Names.csv")
names = pd.read_csv("https://raw.githubusercontent.com/hydro9811/Name-Comparison-/main/name.csv")

In [11]:
df2 = df

Data Cleaning

In [7]:
# id is all the original names from census and title is the known name to compare
id = df.name.map(str)
title = names.title.map(str)

In [8]:
df['dob'] = df['dob'].fillna(0)
df['dob'] = df.dob.map(int)

In [9]:
id2 = [df.name.iloc[i] for i in range(len(df.name))]
title2 = [names.title.iloc[i] for i in range(len(names.title))]

Now group by the origin sources by the language countries

In [10]:
gy = names.groupby('used_in')
gy.size()

used_in
African and English speaking countries                                                                                                                               9
African, Arabic and English speaking countries                                                                                                                       5
African, Arabic, English and Swahili speaking countries                                                                                                              8
African, Arabic, English, Hebrew and Swahili speaking countries                                                                                                      1
African, Arabic, English, Hindi, Hungarian and Swahili speaking countries                                                                                            1
African, Arabic, English, Swahili and Russian speaking countries                                                                                             

In [14]:
groups = gy.describe()
groups = list(groups.index)

Now focus on the languages catagories

In [16]:
African_English = gy.get_group('African and English speaking countries').title.map(str)
African_Arabic_English = gy.get_group('African, Arabic and English speaking countries').title.map(str)
African_Arabic_English_Swahili = gy.get_group(
    'African, Arabic, English and Swahili speaking countries').title.map(str)
African_Arabic_English_Hebrew_Swahili = gy.get_group(
    'African, Arabic, English, Hebrew and Swahili speaking countries').title.map(str)
African_Arabic_English_Hindi_Hungarian_Swahili = gy.get_group(
    'African, Arabic, English, Hindi, Hungarian and Swahili speaking countries').title.map(str)
African_Arabic_English_Swahili_Russian = gy.get_group(
    'African, Arabic, English, Swahili and Russian speaking countries').title.map(str)
African_Bulgarian_Croatian_Czech_English_Polish_Serbian_Spanish_Swahili_Russian = gy.get_group(
    'African, Bulgarian, Croatian, Czech, English, Polish, Serbian, Spanish, Swahili and Russian speaking countries').title.map(str)
African_English_Swahili = gy.get_group('African, English and Swahili speaking countries').title.map(str)
African_English_German_Swahili = gy.get_group('African, English, German and Swahili speaking countries').title.map(str)
African_English_Hebrew_Polish = gy.get_group('African, English, Hebrew and Polish speaking countries').title.map(str)
African_English_Hebrew_Swahili = gy.get_group('African, English, Hebrew and Swahili speaking countries').title.map(str)
African_English_Hindi_Swahili = gy.get_group('African, English, Hindi and Swahili speaking countries').title.map(str)
Albanian_English_Greek_Russian = gy.get_group('Albanian, English, Greek and Russian speaking countries').title
Arabic_English = gy.get_group('Arabic and English speaking countries').title
Arabic_Czech_English_Slovak = gy.get_group('Arabic, Czech, English and Slovak speaking countries').title
Arabic_Czech_English_Hebrew_Hindi_Polish_Slovak = gy.get_group(
    'Arabic, Czech, English, Hebrew, Hindi, Polish and Slovak speaking countries').title
Arabic_English_Greek = gy.get_group('Arabic, English and Greek speaking countries').title
Arabic_English_Hebrew = gy.get_group('Arabic, English and Hebrew speaking countries').title
Arabic_English_Hindi = gy.get_group('Arabic, English and Hindi speaking countries').title
Arabic_English_Japanese = gy.get_group('Arabic, English and Japanese speaking countries').title
Arabic_English_Persian = gy.get_group('Arabic, English and Persian speaking countries').title
Arabic_English_Polish = gy.get_group('Arabic, English and Polish speaking countries').title
Arabic_English_Swahili = gy.get_group('Arabic, English and Swahili speaking countries').title
Arabic_English_French_Greek_Armenian = gy.get_group(
    'Arabic, English, French, Greek and Armenian speaking countries').title
Arabic_English_Hindi_Spanish = gy.get_group('Arabic, English, Hindi and Spanish speaking countries').title
Arabic_English_Portuguese_Spanish_Swahili = gy.get_group(
    'Arabic, English, Portuguese, Spanish and Swahili speaking countries').title
Bulgarian_English = gy.get_group('Bulgarian and English speaking countries').title
Bulgarian_Catalan_Czech_Dutch_English_Estonian_Finnish_German_Greek_Hindi_Hungarian_Italian_Japanese_Polish_Slovak_Armenian = gy.get_group(
    'Bulgarian, Catalan, Czech, Dutch, English, Estonian, Finnish, German, Greek, Hindi, Hungarian, Italian, Japanese, Polish, Slovak and Armenian speaking countries').title
Bulgarian_Croatian_Czech_Dutch_English_Hungarian_Polish_Romanian_Serbian = gy.get_group(
    'Bulgarian, Croatian, Czech, Dutch, English, Hungarian, Polish, Romanian and Serbian speaking countries').title
Bulgarian_Croatian_English_Hindi_Portuguese_Romanian_Serbian_Spanish = gy.get_group(
    'Bulgarian, Croatian, English, Hindi, Portuguese, Romanian, Serbian and Spanish speaking countries').title
Bulgarian_Croatian_English_Hungarian_Polish_Portuguese_Serbian_Russian_Armenian = gy.get_group(groups[30]).title
Bulgarian_Czech_Dutch_English_German_Polish_Romanian_Spanish = gy.get_group(groups[31]).title
Bulgarian_Czech_Dutch_English_German_Romanian_Russian = gy.get_group(groups[32]).title
Bulgarian_Czech_English_Russian = gy.get_group(groups[33]).title
Bulgarian_Czech_English_Polish_Serbian_Slovak_Russian = gy.get_group(groups[34]).title
Bulgarian_English_Hebrew = gy.get_group(groups[35]).title
Bulgarian_English_Russian = gy.get_group(groups[36]).title
Bulgarian_English_Spanish = gy.get_group(groups[37]).title
Bulgarian_English_German_Greek_Italian_Polish_Portuguese_Serbian_Spanish_Russian = gy.get_group(groups[38]).title
Bulgarian_English_Greek_Lithuanian_Spanish_Russian = gy.get_group(groups[39]).title
Catalan_English = gy.get_group(groups[40]).title
Catalan_Dutch_English_German = gy.get_group(groups[41]).title
Catalan_Dutch_English_Estonian_German_Greek_Hungarian_Italian_Norwegian_Polish_Portuguese_Slovak_Spanish_Swedish_Armenian = gy.get_group(groups[42]).title
Catalan_English_French_Hungarian_Norwegian_Polish_Spanish_Swedish_Armenian = gy.get_group(groups[43]).title
Catalan_English_French_Spanish_Swedish = gy.get_group(groups[44]).title
Chinese_English = gy.get_group(groups[45]).title
Croatian_English_Serbian = gy.get_group(groups[46]).title
Croatian_English_German_Hungarian_Lithuanian_Serbian_Swedish = gy.get_group(groups[47]).title
Czech_English = gy.get_group(groups[48]).title
Czech_Danish_Dutch_English_French_German_Hungarian_Norwegian_Swedish = gy.get_group(groups[49]).title
Czech_Danish_English_German = gy.get_group(groups[50]).title
Czech_Danish_English_German_Hungarian_Norwegian_Slovak_Swedish = gy.get_group(groups[51]).title
Czech_Dutch_English_French_German = gy.get_group(groups[52]).title
Czech_Dutch_English_French_German_Hebrew_Polish_Portuguese_Romanian_Spanish_Armenian = gy.get_group(groups[53]).title
Czech_Dutch_English_German_Polish = gy.get_group(groups[54]).title
Czech_Dutch_English_German_Greek_Hungarian_Portuguese_Romanian_Slovak_Russian = gy.get_group(groups[55]).title
Czech_English_French_German = gy.get_group(groups[56]).title
Czech_English_German_Hungarian_Polish = gy.get_group(groups[57]).title
Czech_English_German_Hungarian_Polish_Romanian_Slovak = gy.get_group(groups[58]).title
Czech_English_German_Italian_Romanian_Spanish = gy.get_group(groups[59]).title
Czech_English_German_Latvian_Lithuanian_Norwegian_Polish = gy.get_group(groups[60]).title
Czech_English_Greek_Russian = gy.get_group(groups[61]).title
Czech_English_Hungarian_Japanese_Slovak = gy.get_group(groups[62]).title
Czech_English_Polish_Russian = gy.get_group(groups[63]).title
Czech_English_Polish_Romanian_Spanish = gy.get_group(groups[64]).title
Danish_English = gy.get_group(groups[65]).title
Danish_Dutch_English_French_Swedish = gy.get_group(groups[66]).title
Danish_Dutch_English_French_German_Norwegian_Swedish = gy.get_group(groups[67]).title
Danish_Dutch_English_German_Norwegian = gy.get_group(groups[68]).title
Danish_Dutch_English_Norwegian_Swedish = gy.get_group(groups[69]).title
Danish_English_French = gy.get_group(groups[70]).title
Danish_English_German = gy.get_group(groups[71]).title
Danish_English_Norwegian = gy.get_group(groups[72]).title
Danish_English_Swedish = gy.get_group(groups[73]).title
Danish_English_French_Norwegian = gy.get_group(groups[74]).title
Danish_English_German_Italian_Norwegian_Portuguese_Spanish = gy.get_group(groups[75]).title
Danish_English_German_Norwegian_Spanish = gy.get_group(groups[76]).title
Danish_English_German_Norwegian_Swedish = gy.get_group(groups[77]).title
Danish_English_Hindi_Spanish = gy.get_group(groups[78]).title
Danish_English_Italian_Portuguese_Romanian_Spanish_Swedish = gy.get_group(groups[79]).title
Dutch_English = gy.get_group(groups[80]).title
Dutch_English_French = gy.get_group(groups[81]).title
Dutch_English_German = gy.get_group(groups[82]).title
Dutch_English_Hebrew = gy.get_group(groups[83]).title
Dutch_English_Polish = gy.get_group(groups[84]).title
Dutch_English_Swedish = gy.get_group(groups[85]).title
Dutch_English_Estonian_German_Spanish_Armenian = gy.get_group(groups[86]).title
Dutch_English_Finnish_Swedish = gy.get_group(groups[87]).title
Dutch_English_Finnish_German_Japanese_Swedish = gy.get_group(groups[88]).title
Dutch_English_Finnish_Hindi_Swedish = gy.get_group(groups[89]).title
Dutch_English_French_German = gy.get_group(groups[90]).title
Dutch_English_French_Spanish = gy.get_group(groups[91]).title
Dutch_English_French_German_Norwegian = gy.get_group(groups[92]).title
Dutch_English_French_German_Portuguese = gy.get_group(groups[93]).title
Dutch_English_French_German_Swahili = gy.get_group(groups[94]).title
Dutch_English_French_Persian_Spanish = gy.get_group(groups[95]).title
Dutch_English_German_Hungarian = gy.get_group(groups[96]).title
Dutch_English_German_Polish = gy.get_group(groups[97]).title
Dutch_English_German_Hindi_Italian_Norwegian_Swedish = gy.get_group(groups[98]).title
Dutch_English_German_Hungarian_Slovak = gy.get_group(groups[99]).title
Dutch_English_German_Italian_Polish_Spanish = gy.get_group(groups[100]).title
Dutch_English_German_Norwegian_Swedish = gy.get_group(groups[101]).title
Dutch_English_German_Portuguese_Spanish = gy.get_group(groups[102]).title
Dutch_English_Italian_Spanish = gy.get_group(groups[103]).title
Dutch_English_Italian_Portuguese_Spanish = gy.get_group(groups[104]).title
English_Armenian = gy.get_group(groups[105]).title
English_Estonian = gy.get_group(groups[106]).title
English_Finnish = gy.get_group(groups[107]).title
English_French = gy.get_group(groups[108]).title
English_Gaelic = gy.get_group(groups[109]).title
English_German = gy.get_group(groups[110]).title
English_Greek = gy.get_group(groups[111]).title
English_Hawaiian = gy.get_group(groups[112]).title
English_Hebrew = gy.get_group(groups[113]).title
English_Hindi = gy.get_group(groups[114]).title
English_Hungarian = gy.get_group(groups[115]).title
English_Indonesian = gy.get_group(groups[116]).title
English_Italian = gy.get_group(groups[117]).title
English_Japanese = gy.get_group(groups[118]).title
English_Norwegian = gy.get_group(groups[119]).title
English_Persian = gy.get_group(groups[120]).title
English_Polish = gy.get_group(groups[121]).title
English_Portuguese = gy.get_group(groups[122]).title
English_Romanian = gy.get_group(groups[123]).title
English_Russian = gy.get_group(groups[124]).title
English_Slovak = gy.get_group(groups[125]).title
English_Spanish = gy.get_group(groups[126]).title
English_Swahili = gy.get_group(groups[127]).title
English_Swedish = gy.get_group(groups[128]).title
English_Turkish = gy.get_group(groups[129]).title
English_Urdu = gy.get_group(groups[130]).title
English_Vietnamese = gy.get_group(groups[131]).title
English_Welsh = gy.get_group(groups[132]).title
English_Yiddish = gy.get_group(groups[133]).title
English = gy.get_group(groups[134]).title
English_Estonian_German = gy.get_group(groups[135]).title
English_Finnish_German = gy.get_group(groups[136]).title
English_Finnish_Hindi = gy.get_group(groups[137]).title
English_Finnish_Japanese = gy.get_group(groups[138]).title
English_Finnish_Spanish = gy.get_group(groups[139]).title
English_Finnish_German_Italian_Portuguese_Spanish = gy.get_group(groups[140]).title
English_Finnish_German_Spanish_Swedish = gy.get_group(groups[141]).title
English_Finnish_Hindi_Italian_Spanish = gy.get_group(groups[142]).title
English_Finnish_Polish_Portuguese_Swedish = gy.get_group(groups[143]).title
English_French_Armenian = gy.get_group(groups[144]).title
English_French_German = gy.get_group(groups[145]).title
English_French_Greek = gy.get_group(groups[146]).title
English_French_Italian = gy.get_group(groups[147]).title
English_French_Norwegian = gy.get_group(groups[148]).title
English_French_Polish = gy.get_group(groups[149]).title
English_French_Spanish = gy.get_group(groups[150]).title
English_French_Swedish = gy.get_group(groups[151]).title
English_French_Welsh = gy.get_group(groups[152]).title
English_French_German_Greek = gy.get_group(groups[153]).title
English_French_German_Italian = gy.get_group(groups[154]).title
English_French_German_Polish = gy.get_group(groups[155]).title
English_French_German_Hebrew_Norwegian_Swedish = gy.get_group(groups[156]).title
English_French_German_Polish_Russian_Armenian = gy.get_group(groups[157]).title
English_French_Greek_Italian_Polish = gy.get_group(groups[158]).title
English_French_Greek_Italian_Polish_Portuguese_Spanish = gy.get_group(groups[159]).title
English_French_Hebrew_Japanese = gy.get_group(groups[160]).title
English_French_Hindi_Japanese_Spanish = gy.get_group(groups[161]).title
English_French_Portuguese_Spanish = gy.get_group(groups[162]).title
English_French_Spanish_Swedish = gy.get_group(groups[163]).title
English_Gaelic_Swedish = gy.get_group(groups[164]).title
English_German_Greek = gy.get_group(groups[165]).title
English_German_Hebrew = gy.get_group(groups[166]).title
English_German_Hindi = gy.get_group(groups[167]).title
English_German_Hungarian = gy.get_group(groups[168]).title
English_German_Italian = gy.get_group(groups[169]).title
English_German_Polish = gy.get_group(groups[170]).title
English_German_Spanish = gy.get_group(groups[171]).title
English_German_Greek_Italian = gy.get_group(groups[172]).title
English_German_Greek_Spanish = gy.get_group(groups[173]).title
English_German_Hungarian_Armenian = gy.get_group(groups[174]).title
English_German_Italian_Spanish = gy.get_group(groups[175]).title
English_German_Italian_Polish_Russian = gy.get_group(groups[176]).title
English_Greek_Hebrew = gy.get_group(groups[177]).title
English_Greek_Spanish = gy.get_group(groups[178]).title
English_Hebrew_Armenian = gy.get_group(groups[179]).title
English_Hebrew_Hindi = gy.get_group(groups[180]).title
English_Hebrew_Japanese = gy.get_group(groups[181]).title
English_Hebrew_Polish = gy.get_group(groups[182]).title
English_Hebrew_Spanish = gy.get_group(groups[183]).title
English_Hebrew_Hindi_Hungarian = gy.get_group(groups[184]).title
English_Hindi_Hawaiian = gy.get_group(groups[185]).title
English_Hindi_Indonesian = gy.get_group(groups[186]).title
English_Hindi_Russian = gy.get_group(groups[187]).title
English_Hindi_Welsh = gy.get_group(groups[188]).title
English_Hungarian_Italian = gy.get_group(groups[189]).title
English_Hungarian_Polish = gy.get_group(groups[190]).title
English_Hungarian_Italian_Polish_Portuguese_Romanian_Spanish_Armenian = gy.get_group(groups[191]).title
English_Indonesian_Polish = gy.get_group(groups[192]).title
English_Italian_Polish = gy.get_group(groups[193]).title
English_Italian_Portuguese = gy.get_group(groups[194]).title
English_Italian_Spanish = gy.get_group(groups[195]).title
English_Italian_Polish_Spanish = gy.get_group(groups[196]).title
English_Italian_Polish_Spanish_Russian = gy.get_group(groups[197]).title
English_Italian_Portuguese_Spanish = gy.get_group(groups[198]).title
English_Italian_Portuguese_Swedish = gy.get_group(groups[199]).title
English_Italian_Portuguese_Romanian_Spanish = gy.get_group(groups[200]).title
English_Italian_Portuguese_Serbian_Spanish = gy.get_group(groups[201]).title
English_Italian_Spanish_Swedish = gy.get_group(groups[202]).title
English_Japanese_Welsh = gy.get_group(groups[203]).title
English_Japanese_Yiddish = gy.get_group(groups[204]).title
English_Japanese_Vietnamese_Welsh = gy.get_group(groups[205]).title
English_Latvian_Lithuanian = gy.get_group(groups[206]).title
English_Norwegian_Icelandic = gy.get_group(groups[207]).title
English_Norwegian_Swedish = gy.get_group(groups[208]).title
English_Norwegian_Polish_Swedish = gy.get_group(groups[209]).title
English_Norwegian_Spanish_Swedish = gy.get_group(groups[210]).title
English_Polish_Slovak = gy.get_group(groups[211]).title
English_Polish_Spanish = gy.get_group(groups[212]).title
English_Polish_Spanish_Russian = gy.get_group(groups[213]).title
English_Polish_Swahili_Russian = gy.get_group(groups[214]).title
English_Portuguese_Spanish = gy.get_group(groups[215]).title
English_Portuguese_Spanish_Welsh = gy.get_group(groups[216]).title
English_Romanian_Spanish = gy.get_group(groups[217]).title
English_Russian_Armenian = gy.get_group(groups[218]).title
English_Spanish_Swedish = gy.get_group(groups[219]).title
English_Ukrainian_Russian = gy.get_group(groups[220]).title

Now summarize the languages. For example, African_Arabic_English counted as african, Arabic_Englis counted as arabic and never count same name region catagory twice. 

In [17]:
african = African_Arabic_English.append(African_Arabic_English_Hebrew_Swahili).append(
    African_Arabic_English_Hindi_Hungarian_Swahili).append(African_Arabic_English_Swahili).append(
    African_Arabic_English_Swahili_Russian).append(
    African_Bulgarian_Croatian_Czech_English_Polish_Serbian_Spanish_Swahili_Russian).append(African_English).append(
    African_English_German_Swahili).append(African_English_Hebrew_Polish).append(African_English_Hebrew_Swahili)

albanian = Albanian_English_Greek_Russian

arabic = Arabic_Czech_English_Hebrew_Hindi_Polish_Slovak.append(Arabic_Czech_English_Slovak).append(
    Arabic_English).append(Arabic_English_French_Greek_Armenian).append(Arabic_English_Greek).append(
    Arabic_English_Hebrew).append(Arabic_English_Hindi).append(Arabic_English_Hindi_Spanish).append(
    Arabic_English_Japanese).append(Arabic_English_Persian).append(Arabic_English_Polish).append(
    Arabic_English_Portuguese_Spanish_Swahili).append(Arabic_English_Swahili)

bulgarian = Bulgarian_Catalan_Czech_Dutch_English_Estonian_Finnish_German_Greek_Hindi_Hungarian_Italian_Japanese_Polish_Slovak_Armenian.append(
    Bulgarian_Croatian_Czech_Dutch_English_Hungarian_Polish_Romanian_Serbian).append(
    Bulgarian_Croatian_English_Hindi_Portuguese_Romanian_Serbian_Spanish).append(
    Bulgarian_Croatian_English_Hungarian_Polish_Portuguese_Serbian_Russian_Armenian).append(
    Bulgarian_Czech_Dutch_English_German_Polish_Romanian_Spanish).append(
    Bulgarian_Czech_Dutch_English_German_Romanian_Russian).append(
    Bulgarian_Czech_English_Polish_Serbian_Slovak_Russian).append(Bulgarian_Czech_English_Russian).append(
    Bulgarian_English).append(Bulgarian_English_German_Greek_Italian_Polish_Portuguese_Serbian_Spanish_Russian).append(
    Bulgarian_English_Greek_Lithuanian_Spanish_Russian).append(Bulgarian_English_Hebrew).append(
    Bulgarian_English_Russian).append(Bulgarian_English_Spanish)

catalan = Catalan_Dutch_English_Estonian_German_Greek_Hungarian_Italian_Norwegian_Polish_Portuguese_Slovak_Spanish_Swedish_Armenian.append(
    Catalan_Dutch_English_German).append(Catalan_English).append(
    Catalan_English_French_Hungarian_Norwegian_Polish_Spanish_Swedish_Armenian).append(
    Catalan_English_French_Spanish_Swedish)

chinese = Chinese_English

croatian = Croatian_English_German_Hungarian_Lithuanian_Serbian_Swedish.append(Croatian_English_Serbian)

czech = Czech_Danish_Dutch_English_French_German_Hungarian_Norwegian_Swedish.append(Czech_Danish_English_German).append(
    Czech_Danish_English_German_Hungarian_Norwegian_Slovak_Swedish).append(Czech_Dutch_English_French_German).append(
    Czech_Dutch_English_French_German_Hebrew_Polish_Portuguese_Romanian_Spanish_Armenian).append(
    Czech_Dutch_English_German_Greek_Hungarian_Portuguese_Romanian_Slovak_Russian).append(
    Czech_Dutch_English_German_Polish).append(Czech_English).append(Czech_English_French_German).append(
    Czech_English_German_Hungarian_Polish).append(Czech_English_German_Hungarian_Polish_Romanian_Slovak).append(
    Czech_English_German_Italian_Romanian_Spanish).append(Czech_English_German_Latvian_Lithuanian_Norwegian_Polish).append(
    Czech_English_Greek_Russian).append(Czech_English_Hungarian_Japanese_Slovak).append(
    Czech_English_Polish_Romanian_Spanish).append(Czech_English_Polish_Russian)

danish = Danish_Dutch_English_French_German_Norwegian_Swedish.append(Danish_Dutch_English_French_Swedish).append(
    Danish_Dutch_English_German_Norwegian).append(Danish_Dutch_English_Norwegian_Swedish).append(Danish_English).append(
    Danish_English_French).append(Danish_English_French_Norwegian).append(Danish_English_German).append(
    Danish_English_German_Italian_Norwegian_Portuguese_Spanish).append(Danish_English_German_Norwegian_Spanish).append(
    Danish_English_German_Norwegian_Swedish).append(Danish_English_Hindi_Spanish).append(
    Danish_English_Italian_Portuguese_Romanian_Spanish_Swedish).append(Danish_English_Norwegian).append(
    Danish_English_Swedish)

dutch = Dutch_English.append(Dutch_English_Estonian_German_Spanish_Armenian).append(
    Dutch_English_Finnish_German_Japanese_Swedish).append(Dutch_English_Finnish_Hindi_Swedish).append(
    Dutch_English_Finnish_Swedish).append(Dutch_English_French).append(Dutch_English_French_German).append(
    Dutch_English_French_German_Norwegian).append(Dutch_English_French_German_Portuguese).append(
    Dutch_English_French_German_Swahili).append(Dutch_English_French_Persian_Spanish).append(
    Dutch_English_French_Spanish).append(Dutch_English_German).append(
    Dutch_English_German_Hindi_Italian_Norwegian_Swedish).append(Dutch_English_German_Hungarian).append(
    Dutch_English_German_Hungarian_Slovak).append(Dutch_English_German_Italian_Polish_Spanish).append(
    Dutch_English_German_Norwegian_Swedish).append(Dutch_English_German_Polish).append(
    Dutch_English_German_Portuguese_Spanish).append(Dutch_English_Hebrew).append(
    Dutch_English_Italian_Portuguese_Spanish).append(Dutch_English_Italian_Spanish).append(Dutch_English_Polish).append(
    Dutch_English_Polish)

armenian = English_Armenian

estonian = English_Estonian.append(English_Estonian_German)

finnish = English_Finnish.append(English_Finnish_German).append(English_Finnish_German_Italian_Portuguese_Spanish).append(
    English_Finnish_German_Spanish_Swedish).append(English_Finnish_Hindi).append(English_Finnish_Hindi_Italian_Spanish).append(
    English_Finnish_Japanese).append(English_Finnish_Polish_Portuguese_Swedish).append(English_Finnish_Spanish)

french = English_French.append(English_French_Armenian).append(English_French_German).append(
    English_French_German_Greek).append(English_French_German_Hebrew_Norwegian_Swedish).append(
    English_French_German_Italian).append(English_French_German_Polish).append(
    English_French_German_Polish_Russian_Armenian).append(English_French_Greek).append(
    English_French_Greek_Italian_Polish).append(English_French_Greek_Italian_Polish_Portuguese_Spanish).append(
    English_French_Hebrew_Japanese).append(English_French_Hindi_Japanese_Spanish).append(English_French_Italian).append(
    English_French_Norwegian).append(English_French_Polish).append(English_French_Portuguese_Spanish).append(
    English_French_Spanish).append(English_French_Spanish_Swedish).append(English_French_Swedish).append(
    English_French_Welsh)

gaelic = English_Gaelic.append(English_Gaelic_Swedish)

german = English_German.append(English_German_Greek).append(English_German_Greek_Italian).append(
    English_German_Greek_Spanish).append(English_German_Hebrew).append(English_German_Hindi).append(
    English_German_Hungarian).append(English_German_Hungarian_Armenian).append(English_German_Italian).append(
    English_German_Italian_Polish_Russian).append(English_German_Italian_Spanish).append(English_German_Polish).append(
    English_German_Spanish)

greek = English_Greek.append(English_Greek_Hebrew).append(English_Greek_Spanish)

hawaiian = English_Hawaiian

hebrew = English_Hebrew.append(English_Hebrew_Armenian).append(English_Hebrew_Hindi).append(
    English_Hebrew_Hindi_Hungarian).append(English_Hebrew_Japanese).append(English_Hebrew_Polish).append(
    English_Hebrew_Spanish)

hindi = English_Hindi.append(English_Hindi_Hawaiian).append(English_Hindi_Indonesian).append(English_Hindi_Russian).append(
    English_Hindi_Welsh)

hungarian = English_Hungarian.append(English_Hungarian_Italian).append(
    English_Hungarian_Italian_Polish_Portuguese_Romanian_Spanish_Armenian).append(English_Hungarian_Polish)

indonesian = English_Indonesian.append(English_Indonesian_Polish)

italian = English_Italian.append(English_Italian_Polish).append(English_Italian_Polish_Spanish).append(
    English_Italian_Polish_Spanish_Russian).append(English_Italian_Portuguese).append(
    English_Italian_Portuguese_Romanian_Spanish).append(English_Italian_Portuguese_Serbian_Spanish).append(
    English_Italian_Portuguese_Spanish).append(English_Italian_Portuguese_Swedish).append(
    English_Italian_Spanish).append(English_Italian_Spanish_Swedish)

japanese = English_Japanese.append(English_Japanese_Vietnamese_Welsh).append(English_Japanese_Welsh).append(
    English_Japanese_Yiddish)

norwegian = English_Norwegian.append(English_Norwegian_Icelandic).append(English_Norwegian_Polish_Swedish).append(
    English_Norwegian_Spanish_Swedish).append(English_Norwegian_Swedish)

persian = English_Persian

polish = English_Polish.append(English_Polish_Slovak).append(English_Polish_Spanish).append(
    English_Polish_Spanish_Russian).append(English_Polish_Swahili_Russian)

portuguese = English_Portuguese.append(English_Portuguese_Spanish).append(English_Portuguese_Spanish_Welsh)

romanian = English_Romanian.append(English_Romanian_Spanish)

russian = English_Russian.append(English_Russian_Armenian)

slovak = English_Slovak

spanish = English_Spanish.append(English_Spanish_Swedish)

swahili = English_Swahili

swedish = English_Swedish

turkish = English_Turkish

urdu = English_Urdu

vietnamese = English_Vietnamese

welsh = English_Welsh

yiddish = English_Yiddish

english = English

latvian = English_Latvian_Lithuanian

ukrainian = English_Ukrainian_Russian

In [18]:
df2 = df2.drop(['1 for Euro Name: Original', 'European Name'], axis=1)

In [19]:
#42 catagories
l = ['african', 'albanian', 'arabic', 'bulgarian', 'catalan', 'chinese', 'croatian', 'czech', 'danish', 'dutch'
     , 'armenian', 'estonian', 'finnish', 'french', 'gaelic', 'german', 'greek', 'hawaiian', 'hebrew', 'hindi'
     , 'hungarian', 'indonesian', 'italian', 'japanese', 'norwegian', 'persian', 'polish', 'portuguese', 'romanian'
     , 'russian', 'slovak', 'spanish', 'swahili', 'swedish', 'turkish', 'urdu', 'vietnamese', 'welsh', 'yiddish'
     , 'english', 'latvian', 'ukrainian']
for col in l:
    df2[col] = 0

In [20]:
df2.head()

ID                 name  \
0   82336         Canngegasick   
1   84111            Casbellos   
2  176251   Maymahakawahcuming   
3  209123                Rocks   
4  228206        Stands On Top   

                                                city   dob  african  albanian  \
0                      Norwood, Charlevoix, Michigan  1859        0         0   
1            San Juan Pueblo, Rio Arriba, New Mexico  1848        0         0   
2  White Earth Indian Reservation, Becker, Minnesota  1888        0         0   
3                 Smith River, Del Norte, California  1893        0         0   
4  Crow Creek Indian Reservation, Buffalo, South ...  1864        0         0   

   arabic  bulgarian  catalan  chinese  ...  swahili  swedish  turkish  urdu  \
0       0          0        0        0  ...        0        0        0     0   
1       0          0        0        0  ...        0        0        0     0   
2       0          0        0        0  ...        0        0        0     0   
3       0          0        0        0  ...        0        0        0     0   
4       0          0        0        0  ...        0        0        0     0   

   vietnamese  welsh  yiddish  english  latvian  ukrainian  
0           0      0        0        0        0          0  
1           0      0        0        0        0          0  
2           0      0        0        0        0          0  
3           0      0        0        0        0          0  
4           0      0        0        0        0          0  

[5 rows x 46 columns]

In [21]:
df3 = df2

Start compare names in census data to the known name data

In [15]:
#0 to 9999 
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[i], african[african.index[j]])
        if x > df3['african'][i]:
            df3['african'][i] = x

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [21]:
#10000 to 19999 Takes 39 minutes to run this cell code
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[10000+i], african[african.index[j]])
        if x > df3['african'][10000+i]:
            df3['african'][10000+i] = x

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [24]:
#20000 to 29999 
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[20000+i], african[african.index[j]])
        if x > df3['african'][20000+i]:
            df3['african'][20000+i] = x

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [26]:
#30000 to 39999 
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[30000+i], african[african.index[j]])
        if x > df3['african'][30000+i]:
            df3['african'][30000+i] = x

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [29]:
#40000 to 49999 This cell took me over 55 mins to run it
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[40000+i], african[african.index[j]])
        if x > df3['african'][40000+i]:
            df3['african'][40000+i] = x

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [31]:
#50000 to 59999 (50mins)
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[50000+i], african[african.index[j]])
        if x > df3['african'][50000+i]:
            df3['african'][50000+i] = x

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [33]:
#60000 to 69999 
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[60000+i], african[african.index[j]])
        if x > df3['african'][60000+i]:
            df3['african'][60000+i] = x

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [40]:
#70000 to 79999 
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[70000+i], african[african.index[j]])
        if x > df3['african'][70000+i]:
            df3['african'][70000+i] = x

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


In [44]:
#80000 to 89999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[80000+i], african[african.index[j]])
        if x > df3['african'][80000+i]:
            df3['african'][80000+i] = x
print(datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


0:44:19.347847


In [46]:
#90000 to 99999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[90000+i], african[african.index[j]])
        if x > df3['african'][90000+i]:
            df3['african'][90000+i] = x
print(datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


0:56:42.975020


In [54]:
#100000 to 109999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[100000+i], african[african.index[j]])
        if x > df3['african'][100000+i]:
            df3['african'][100000+i] = x
print('running time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


running time used: 0:58:50.327537


In [60]:
#110000 to 119999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[110000+i], african[african.index[j]])
        if x > df3['african'][110000+i]:
            df3['african'][110000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


running time used: 1:08:47.416292


In [62]:
#120000 to 129999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[120000+i], african[african.index[j]])
        if x > df3['african'][120000+i]:
            df3['african'][120000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 1:04:49.259037


In [63]:
#130000 to 139999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[130000+i], african[african.index[j]])
        if x > df3['african'][130000+i]:
            df3['african'][130000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 1:06:31.477480


In [69]:
#140000 to 149999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[140000+i], african[african.index[j]])
        if x > df3['african'][140000+i]:
            df3['african'][140000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 1:06:17.944345


In [70]:
#150000 to 159999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[150000+i], african[african.index[j]])
        if x > df3['african'][150000+i]:
            df3['african'][150000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 0:55:26.140294


In [71]:
#160000 to 169999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[160000+i], african[african.index[j]])
        if x > df3['african'][160000+i]:
            df3['african'][160000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 0:23:24.811052


In [72]:
#170000 to 179999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[170000+i], african[african.index[j]])
        if x > df3['african'][170000+i]:
            df3['african'][170000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 0:55:31.745706


In [73]:
#180000 to 189999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[180000+i], african[african.index[j]])
        if x > df3['african'][180000+i]:
            df3['african'][180000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 1:31:59.377666


In [74]:
#190000 to 199999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[190000+i], african[african.index[j]])
        if x > df3['african'][190000+i]:
            df3['african'][190000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 1:00:52.188613


In [75]:
#200000 to 209999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[200000+i], african[african.index[j]])
        if x > df3['african'][200000+i]:
            df3['african'][200000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 0:44:53.073124


In [76]:
#210000 to 219999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[210000+i], african[african.index[j]])
        if x > df3['african'][210000+i]:
            df3['african'][210000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 0:56:21.926043


In [77]:
#220000 to 229999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[220000+i], african[african.index[j]])
        if x > df3['african'][220000+i]:
            df3['african'][220000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 0:48:16.165676


In [78]:
#230000 to 239999 
begin_time = datetime.datetime.now()
for i in range(10000):
    for j in range(len(african)):
        x = who.ratio(id[230000+i], african[african.index[j]])
        if x > df3['african'][230000+i]:
            df3['african'][230000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

/Users/haizhouli/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


time used: 1:01:29.037730


In [ ]:
#240000 to 245545
begin_time = datetime.datetime.now()
for i in range(5545):
    for j in range(len(african)):
        x = who.ratio(id[240000+i], african[african.index[j]])
        if x > df3['african'][240000+i]:
            df3['african'][240000+i] = x
print('time used:', datetime.datetime.now() - begin_time)

After I compared only afican names, it cost me about 20 hours

In [86]:
gb_african = df3.groupby('african')
gb_african.size()

african
0      3189
3         1
4        23
5        66
6      3637
7      2700
8      2537
9      3844
10     7628
11     2462
12     9875
13    13453
14      871
15    16799
16    27917
17     8948
18    40904
19     1335
20    25107
21     9100
22    19419
23     3556
25    22537
26      147
27     4765
28     4960
30     1463
31      509
32      463
33     1621
34       94
37     5615
dtype: int64

This cell represents the percentage groupby, that 5615 names has 37% chance to be an african name, and 3189 names is impossible to be an african name